framework/decode/dx12_replay_consumer_base.cpp

```diff
@@ -4554,6 +4554,18 @@ void Dx12ReplayConsumerBase::PreCall_ID3D12GraphicsCommandList_ResourceBarrier(
             {
                 it->second.emplace_back(std::move(state));
             }
+            if (options_.enable_dump_resources)
+            {
+                GFXRECON_ASSERT(dump_resources_);
+                if (state.transition.StateAfter & modifiableTransitionStates)
+                {
+                    dump_resources_->ModifiableResourceAdd(barriers[i].Transition->pResource);
+                }
+                else
+                {
+                    dump_resources_->ModifiableResourceRemove(barriers[i].Transition->pResource);
```

Comment:
Here might not be a good place to record states.  `command_list->ResourceBarrier` don't change on the resource states  until the `command_list` is executed on `ExecuteCommandLists`.  A `ResourceBarrier` could be before the target `ExecuteCommandLists`, but its `ExecuteCommandLists` is after the target `ExecuteCommandLists`. In this case, changing state doesn't happen before dump resources.

Getting a correct resource state is complicated. It finds the correct resource states for copying resources on `Dx12DumpResources::CopyResourceAsyncQueue`. Maybe you could decide if the resource should be dumped there.  But from what I remember, it's not prefect. It triggered validations for wrong states sometimes.

@davidd-lunarg Do you think that the wrong states issues still happen sometimes?